File: nautilus_trader/config/components.py
# -------------------------------------------------------------------------------------------------
#  Copyright (C) 2015-2022 Nautech Systems Pty Ltd. All rights reserved.
#  https://nautechsystems.io
#
#  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
#  You may not use this file except in compliance with the License.
#  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
#
#  Unless required by applicable law or agreed to in writing, software
#  distributed under the License is distributed on an "AS IS" BASIS,
#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
#  See the License for the specific language governing permissions and
#  limitations under the License.
# -------------------------------------------------------------------------------------------------

from typing import Any, Dict, FrozenSet, Optional

import pydantic
from frozendict import frozendict
from pydantic import PositiveInt
from pydantic import validator

from nautilus_trader.config.common import resolve_path
from nautilus_trader.core.correctness import PyCondition


class CacheConfig(pydantic.BaseModel):
    """
    Configuration for ``Cache`` instances.

    Parameters
    ----------
    tick_capacity : int
        The maximum length for internal tick deques.
    bar_capacity : int
        The maximum length for internal bar deques.
    """

    tick_capacity: PositiveInt = 1000
    bar_capacity: PositiveInt = 1000


class CacheDatabaseConfig(pydantic.BaseModel):
    """
    Configuration for ``CacheDatabase`` instances.

    Parameters
    ----------
    type : str, {'in-memory', 'redis'}, default 'in-memory'
        The database type.
    host : str, default 'localhost'
        The database host address (default for Redis).
    port : int, default 6379
        The database port (default for Redis).
    flush : bool, default False
        If database should be flushed before start.
    """

    type: str = "in-memory"
    host: str = "localhost"
    port: int = 6379
    flush: bool = False


class ActorConfig(pydantic.BaseModel):
    """
    The base model for all actor configurations.

    Parameters
    ----------
    component_id : str, optional
        The component ID. If ``None`` then the identifier will be taken from
        `type(self).__name__`.

    """

    component_id: Optional[str] = None


class ImportableActorConfig(pydantic.BaseModel):
    """
    Represents an actor configuration for one specific backtest run.

    Parameters
    ----------
    actor_path : str
        The fully qualified name of the Actor class.
    config_path : str
        The fully qualified name of the Actor Config class.
    config : Dict
        The actor configuration
    """

    actor_path: str
    config_path: str
    config: dict


class ActorFactory:
    """
    Provides actor creation from importable configurations.
    """

    @staticmethod
    def create(config: ImportableActorConfig):
        """
        Create an actor from the given configuration.

        Parameters
        ----------
        config : ImportableActorConfig
            The configuration for the building step.

        Returns
        -------
        Actor

        Raises
        ------
        TypeError
            If `config` is not of type `ImportableActorConfig`.

        """
        PyCondition.type(config, ImportableActorConfig, "config")
        strategy_cls = resolve_path(config.actor_path)
        config_cls = resolve_path(config.config_path)
        return strategy_cls(config=config_cls(**config.config))


class TradingStrategyConfig(pydantic.BaseModel):
    """
    The base model for all trading strategy configurations.

    Parameters
    ----------
    strategy_id : str, optional
        The unique ID for the strategy. Will become the strategy ID if not None.
    order_id_tag : str
        The unique order ID tag for the strategy. Must be unique
        amongst all running strategies for a particular trader ID.
    oms_type : OMSType, optional
        The order management system type for the strategy. This will determine
        how the `ExecutionEngine` handles position IDs (see docs).

    """

    strategy_id: Optional[str] = None
    order_id_tag: str = "000"
    oms_type: Optional[str] = None


class ImportableStrategyConfig(pydantic.BaseModel):
    """
    Represents a trading strategy configuration for one specific backtest run.

    Parameters
    ----------
    strategy_path : str
        The fully qualified name of the strategy class.
    config_path : str
        The fully qualified name of the config class.
    config : Dict[str, Any]
        The strategy configuration
    """

    strategy_path: str
    config_path: str
    config: Dict[str, Any]


class StrategyFactory:
    """
    Provides strategy creation from importable configurations.
    """

    @staticmethod
    def create(config: ImportableStrategyConfig):
        """
        Create a trading strategy from the given configuration.

        Parameters
        ----------
        config : ImportableStrategyConfig
            The configuration for the building step.

        Returns
        -------
        TradingStrategy

        Raises
        ------
        TypeError
            If `config` is not of type `ImportableStrategyConfig`.

        """
        PyCondition.type(config, ImportableStrategyConfig, "config")
        strategy_cls = resolve_path(config.strategy_path)
        config_cls = resolve_path(config.config_path)
        return strategy_cls(config=config_cls(**config.config))


class InstrumentProviderConfig(pydantic.BaseModel):
    """
    Configuration for ``InstrumentProvider`` instances.

    Parameters
    ----------
    load_all : bool, default False
        If all venue instruments should be loaded on start.
    load_ids : FrozenSet[str], optional
        The list of instrument IDs to be loaded on start (if `load_all_instruments` is False).
    filters : frozendict, optional
        The venue specific instrument loading filters to apply.
    """

    class Config:
        """The base model config"""

        arbitrary_types_allowed = True

    @validator("filters")
    def validate_filters(cls, value):
        return frozendict(value) if value is not None else None

    def __eq__(self, other):
        return (
            self.load_all == other.load_all
            and self.load_ids == other.load_ids
            and self.filters == other.filters
        )

    def __hash__(self):
        return hash((self.load_all, self.load_ids, self.filters))

    load_all: bool = False
    load_ids: Optional[FrozenSet[str]] = None
    filters: Optional[Dict[str, Any]] = None

File: crates/testkit/src/components.rs
// -------------------------------------------------------------------------------------------------
//  Copyright (C) 2015-2026 Nautech Systems Pty Ltd. All rights reserved.
//  https://nautechsystems.io
//
//  Licensed under the GNU Lesser General Public License Version 3.0 (the "License");
//  You may not use this file except in compliance with the License.
//  You may obtain a copy of the License at https://www.gnu.org/licenses/lgpl-3.0.en.html
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
// -------------------------------------------------------------------------------------------------

//! Stateful actor and strategy test components.

use indexmap::IndexMap;
use nautilus_common::{
    actor::{DataActor, DataActorCore, data_actor::DataActorConfig},
    nautilus_actor,
};
use nautilus_model::identifiers::{ActorId, StrategyId};
use nautilus_trading::{
    nautilus_strategy,
    strategy::{config::StrategyConfig, core::StrategyCore},
};

use crate::cache::TestCacheDatabaseControl;

/// Actor with observable state lifecycle callbacks.
#[derive(Debug)]
pub struct StateActor {
    core: DataActorCore,
    control: TestCacheDatabaseControl,
    state_load: Option<IndexMap<String, Vec<u8>>>,
    state_save: IndexMap<String, Vec<u8>>,
    fail_load: bool,
    fail_save: bool,
    fail_start: bool,
}

impl StateActor {
    /// Creates a stateful actor.
    #[must_use]
    pub fn new(
        actor_id: ActorId,
        control: TestCacheDatabaseControl,
        state_save: IndexMap<String, Vec<u8>>,
    ) -> Self {
        Self {
            core: DataActorCore::new(DataActorConfig {
                actor_id: Some(actor_id),
                ..Default::default()
            }),
            control,
            state_load: None,
            state_save,
            fail_load: false,
            fail_save: false,
            fail_start: false,
        }
    }

    /// Configures the actor load callback to fail.
    #[must_use]
    pub const fn with_fail_load(mut self) -> Self {
        self.fail_load = true;
        self
    }

    /// Configures the actor save callback to fail.
    #[must_use]
    pub const fn with_fail_save(mut self) -> Self {
        self.fail_save = true;
        self
    }

    /// Configures the actor start callback to fail.
    #[must_use]
    pub const fn with_fail_start(mut self) -> Self {
        self.fail_start = true;
        self
    }

    /// Returns the state received by the load callback.
    #[must_use]
    pub const fn state_load(&self) -> Option<&IndexMap<String, Vec<u8>>> {
        self.state_load.as_ref()
    }
}

impl DataActor for StateActor {
    fn on_load(&mut self, state: IndexMap<String, Vec<u8>>) -> anyhow::Result<()> {
        self.control.record("actor.on_load");
        if self.fail_load {
            anyhow::bail!("test actor on_load failure");
        }
        self.state_load = Some(state);
        Ok(())
    }

    fn on_start(&mut self) -> anyhow::Result<()> {
        self.control.record("actor.on_start");
        if self.fail_start {
            anyhow::bail!("test actor on_start failure");
        }
        Ok(())
    }

    fn on_stop(&mut self) -> anyhow::Result<()> {
        self.control.record("actor.on_stop");
        Ok(())
    }

    fn on_save(&self) -> anyhow::Result<IndexMap<String, Vec<u8>>> {
        self.control.record("actor.on_save");
        if self.fail_save {
            anyhow::bail!("test actor on_save failure");
        }
        Ok(self.state_save.clone())
    }
}

nautilus_actor!(StateActor);

/// Strategy with observable state lifecycle callbacks.
#[derive(Debug)]
pub struct StateStrategy {
    core: StrategyCore,
    control: TestCacheDatabaseControl,
    state_load: Option<IndexMap<String, Vec<u8>>>,
    state_save: IndexMap<String, Vec<u8>>,
    fail_load: bool,
    fail_save: bool,
    fail_start: bool,
}

impl StateStrategy {
    /// Creates a stateful strategy.
    #[must_use]
    pub fn new(
        strategy_id: StrategyId,
        control: TestCacheDatabaseControl,
        state_save: IndexMap<String, Vec<u8>>,
    ) -> Self {
        Self {
            core: StrategyCore::new(StrategyConfig {
                strategy_id: Some(strategy_id),
                ..Default::default()
            }),
            control,
            state_load: None,
            state_save,
            fail_load: false,
            fail_save: false,
            fail_start: false,
        }
    }

    /// Configures the strategy load callback to fail.
    #[must_use]
    pub const fn with_fail_load(mut self) -> Self {
        self.fail_load = true;
        self
    }

    /// Configures the strategy save callback to fail.
    #[must_use]
    pub const fn with_fail_save(mut self) -> Self {
        self.fail_save = true;
        self
    }

    /// Configures the strategy start callback to fail.
    #[must_use]
    pub const fn with_fail_start(mut self) -> Self {
        self.fail_start = true;
        self
    }

    /// Returns the state received by the load callback.
    #[must_use]
    pub const fn state_load(&self) -> Option<&IndexMap<String, Vec<u8>>> {
        self.state_load.as_ref()
    }
}

impl DataActor for StateStrategy {
    fn on_load(&mut self, state: IndexMap<String, Vec<u8>>) -> anyhow::Result<()> {
        self.control.record("strategy.on_load");
        if self.fail_load {
            anyhow::bail!("test strategy on_load failure");
        }
        self.state_load = Some(state);
        Ok(())
    }

    fn on_start(&mut self) -> anyhow::Result<()> {
        self.control.record("strategy.on_start");
        if self.fail_start {
            anyhow::bail!("test strategy on_start failure");
        }
        Ok(())
    }

    fn on_stop(&mut self) -> anyhow::Result<()> {
        self.control.record("strategy.on_stop");
        Ok(())
    }

    fn on_save(&self) -> anyhow::Result<IndexMap<String, Vec<u8>>> {
        self.control.record("strategy.on_save");
        if self.fail_save {
            anyhow::bail!("test strategy on_save failure");
        }
        Ok(self.state_save.clone())
    }
}

nautilus_strategy!(StateStrategy);
